storages/csv-storage/src/lib.rs

```diff
@@ -0,0 +1,246 @@
+        fs::create_dir_all(path).map_storage_err()?;
+        let path = PathBuf::from(path);
+
+        Ok(Self { path })
+    }
+
+    fn fetch_schema(&self, table_name: &str) -> Result<Option<(Schema, bool)>> {
+        let schema_path = self.schema_path(table_name);
+        if !schema_path.exists() {
+            let data_path = self.data_path(table_name);
+            let types_path = self.types_path(table_name);
+
+            let column_defs = match (types_path.exists(), data_path.exists()) {
+                (false, false) => return Ok(None),
+                (false, true) => Some(
+                    csv::Reader::from_path(data_path)
+                        .map_storage_err()?
+                        .headers()
+                        .map_storage_err()?
+                        .into_iter()
+                        .map(|header| ColumnDef {
+                            name: header.to_string(),
+                            data_type: DataType::Text,
+                            unique: None,
+                            default: None,
+                            nullable: true,
+                        })
+                        .collect::<Vec<_>>(),
+                ),
+                (true, _) => None,
+            };
+
+            let schema = Schema {
+                table_name: table_name.to_owned(),
+                column_defs,
+                indexes: Vec::new(),
+                engine: None,
+            };
+
+            return Ok(Some((schema, true)));
+        }
+
+        let mut file = File::open(&schema_path).map_storage_err()?;
+        let mut ddl = String::new();
+        file.read_to_string(&mut ddl).map_storage_err()?;
+
+        let schema = Schema::from_ddl(&ddl)?;
+        if schema.table_name != table_name {
+            return Err(CsvStorageError::TableNameDoesNotMatchWithFile.into());
+        }
+
+        Ok(Some((schema, false)))
+    }
+
+    fn path_by(&self, table_name: &str, extension: &str) -> PathBuf {
+        let path = self.path.as_path();
+        let mut path = path.join(table_name);
+        path.set_extension(extension);
+
+        path
+    }
+
+    fn schema_path(&self, table_name: &str) -> PathBuf {
+        self.path_by(table_name, "sql")
+    }
+
+    fn data_path(&self, table_name: &str) -> PathBuf {
+        self.path_by(table_name, "csv")
+    }
+
+    fn types_path(&self, table_name: &str) -> PathBuf {
+        self.path_by(table_name, "types.csv")
+    }
+
+    fn scan_data(&self, table_name: &str) -> Result<(Option<Vec<String>>, RowIter)> {
+        let data_path = self.data_path(table_name);
+        let (schema, generated) = match (self.fetch_schema(table_name)?, data_path.exists()) {
+            (None, _) | (_, false) => return Ok((None, Box::new(std::iter::empty()))),
+            (Some(v), true) => v,
+        };
+
+        let mut data_rdr = csv::Reader::from_path(data_path).map_storage_err()?;
+        let mut fetch_data_header_columns = || -> Result<Vec<String>> {
```

Comment:
Looks like it is okay to use just `immutable` and  `variable`
```rs
let fetch_data_header_columns = data_rdr
    .headers()
    .map_storage_err()?
    .into_iter()
    .map(|header| header.to_string())
    .collect::<Vec<_>>();
```

Reply:
Oh... If it was intended to run only in 2 of 3 conditions, mutable closure would be right.